src/VisualStudio/Core/Def/Implementation/Options/RoamingVisualStudioProfileOptionPersister.cs

```diff
@@ -50,20 +58,33 @@ public RoamingVisualStudioProfileOptionPersister(IThreadingContext threadingCont
         {
             Contract.ThrowIfNull(globalOptionService);
 
-            _settingManager = (ISettingsManager)serviceProvider.GetService(typeof(SVsSettingsPersistenceManager));
             _globalOptionService = globalOptionService;
 
-            // While the settings persistence service should be available in all SKUs it is possible an ISO shell author has undefined the
-            // contributing package. In that case persistence of settings won't work (we don't bother with a backup solution for persistence
-            // as the scenario seems exceedingly unlikely), but we shouldn't crash the IDE.
-            if (_settingManager != null)
-            {
-                var settingsSubset = _settingManager.GetSubset("*");
-                settingsSubset.SettingChangedAsync += OnSettingChangedAsync;
-            }
+            var serviceContainer = serviceProvider.GetService<SVsBrokeredServiceContainer, IBrokeredServiceContainer>();
+            _serviceBroker = serviceContainer.GetFullAccessServiceBroker();
+            _serviceBroker.AvailabilityChanged += Sb_AvailabilityChanged;
```

Comment:
Something that hides the "AvailabilityChanged, let's refetch the service" behind a helper type would be great...the lack of "readonly" is pretty worrisome for those fields.

Reply:
We have the `ServiceBrokerClient` as just such a helper class, but I advise against using it if you add event handlers to whatever service you acquire, since service lifetime is no longer transparent. You *can* make it work (and the VC team does), but it's swimming upstream, IMO.

Reply:
Do you have an example of what the "can make it work" looks like perhaps so we know to recognize the anti-pattern? I'm imagining a pattern where the wrapper would take the delegate to reexecute the event handler addition if necessary. That wouldn't be usable of course for stateful services, but if the only thing is an event handler addition it seems we can do something useful there.

Reply:
The anti-pattern can be recognized by using `ServiceBrokerClient` and *storing* the returned `Rental<T>` struct as a field in your class. The intended consumption is simply this:

```cs
using (var rentedProxy = await serviceBrokerClient.GetProxyAsync<T>(descriptor))
{
   await rentedProxy.Proxy.DoSomethingAsync();
}
```